py/test/selenium/webdriver/common/virtual_authenticator_tests.py

```diff
@@ -144,109 +151,246 @@ def test_add_and_remove_non_resident_credentials(driver, pages):
-    options.protocol = VirtualAuthenticatorOptions.Protocol.U2F
-    options.transport = VirtualAuthenticatorOptions.Transport.USB
+    # register a credential
+    response = driver.execute_async_script(
+        "registerCredential().then(arguments[arguments.length - 1]);"
+    )
+    assert response.get('status', '') == 'OK'
+
+    # remove the credential using array of bytes: rawId
+    raw_id = extract_raw_id(response)
+    driver.remove_credential(bytearray(raw_id))
+
+    # Trying to get the assertion should fail
+    response = get_assertion_for(driver, raw_id)
+    if response.get("status", "").startswith("NotAllowedError"):
+        assert True
+    else:
+        assert False, "Should have thrown a NotAllowedError"
+    driver.remove_virtual_authenticator()
 
-    driver.add_virtual_authenticator(options)
 
-    driver.get("https://webauthn.io/")
-    username = driver.find_element(By.ID, "input-email")
-    username.send_keys("username")
+@pytest.mark.xfail_firefox
+@pytest.mark.xfail_safari
+@pytest.mark.xfail_remote
+def test_remove_credential_by_b64_urlId(driver, pages):
+    driver = create_rk_disabled_u2f_authenticator(driver)
+    driver.get(pages.url("virtual-authenticator.html", localhost=True))
 
-    selectAttestation = Select(driver.find_element(By.ID, "select-attestation"))
-    selectAttestation.select_by_visible_text("Direct")
+    # register a credential
+    response = driver.execute_async_script(
+        "registerCredential().then(arguments[arguments.length - 1]);"
+    )
+    assert response.get('status', '') == 'OK'
+
+    # remove the credential using array of bytes: rawId
+    raw_id = extract_raw_id(response)
+    credential_id = extract_id(response)
+    driver.remove_credential(credential_id)
+
+    # Trying to get the assertion should fail
+    response = get_assertion_for(driver, raw_id)
+    if response.get("status", "").startswith("NotAllowedError"):
+        assert True
+    else:
+        assert False, "Should have thrown a NotAllowedError"
+    driver.remove_virtual_authenticator()
 
-    selectAuthenticator = Select(driver.find_element(By.ID, "select-authenticator"))
-    selectAuthenticator.select_by_value("cross-platform")
 
-    driver.find_element(By.ID, "register-button").click()
+@pytest.mark.xfail_firefox
+@pytest.mark.xfail_safari
+@pytest.mark.xfail_remote
+def test_remove_all_credentials(driver, pages):
+    driver = create_rk_disabled_u2f_authenticator(driver)
 
-    login = driver.find_element(By.ID, "login-button")
-    WebDriverWait(driver, 50).until(lambda x: x.find_element(By.ID, "login-button").is_displayed())
-    login.click()
+    driver.get(pages.url("virtual-authenticator.html", localhost=True))
 
-    WebDriverWait(driver, 40).until(lambda x: x.find_element(By.CLASS_NAME, "col-lg-12").is_displayed())
+    # Register 2 credentials
+    response1 = driver.execute_async_script(
+        "registerCredential().then(arguments[arguments.length - 1]);"
```

Comment:
let's move this JS to a variable and then we can simplify the code here and below. It means any future changes only need to happen in 1 place

Reply:
Sure